### ISSUE TITLE
[BUG] CROS header 'Access-Control-Allow-Origin' is missing

### ISSUE THREAD

**Issue by samsja** [Kong/kubernetes-ingress-controller#411]
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

SUMMARY_GOES_HERE

**Kong Ingress controller version**

kong-ingress 0.6.0

**Kong or Kong Enterprise version**

kong:1.3

### Kubernetes version
```bash
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.7-gke.25", GitCommit:"d4c79083ab6dea5d26ef4ed8d50b145268349bc3", GitTreeState:"clean", BuildDate:"2019-06-22T16:10:31Z", GoVersion:"go1.10.8b4", Compiler:"gc", Platform:"linux/amd64"}
```

### Environment

Gcloud cluster

### What happened

I have two API, one front api and an other api that are working side by side.

I was using previsouly the default google cloud balancer to expose the api.

I decided to switch and use Kong as an API-Gateway because I urgently need rate-limit and black/white list on thoses APIs.

However now that I have replace my loadBanlancer and deploy kong, the APIs can't communicate proprely.

I get the folowing responde from my front-api on chrome :

```error
has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

### Steps To Reproduce
install kong with : 

kubectl apply -f https://bit.ly/kong-ingress

then deploy the ingress

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: "kong"
    nginx.ingress.kubernetes.io/use-regex: "true"
  name: my-ingress
spec:
  rules:
  - host: front-api.my-domain
    http:
      paths:
      - backend:
          serviceName: front-end-service
          servicePort: 80
  - host: back-api.my-domain
    http:
      paths:
      - backend:
          serviceName: back-end-service
          servicePort: 80
```

I have configurate google dns so that my two host adresses points to the kong-proxy ip

### Note

I now that the issues is well known with CROS and kong-ingress, i have tried unsuccefully to use CROS plugin and i am aware of the following solution : https://discuss.konghq.com/t/kong-ingress-controller-with-cors-plugin-failed-with-preflight-options-request/4261

but it won't work becasue i need to acces the the back-end api with the adress back-api.my-domain that point the the kong-proxy butit comes with the Host and won't reach my paths

Thanks

**Comment by hbagdi** [comment]
> but it won't work becasue i need to acces the the back-end api with the adress back-api.my-domain that point the the kong-proxy butit comes with the Host and won't reach my paths

I'm not sure I follow. Could you rephrase?

You can add `Access-Control-Allow-Origin: *` header using the CORS plugin in Kong to resolve the error you are seeing in your browser.

**Comment by samsja** [comment]
Yes i want to use this plugin, but according the this : https://docs.konghq.com/hub/kong-inc/cors/
`Because of this limitation, this plugin will only work for Routes that have been configured with a paths setting, and it will not work for Routes that are being resolved using a custom DNS (the hosts property)`

So I apparently can't use the host preperty in my kong-ingress the redicecrt the url to the right backend service. So how can redirect those request to the kong-proxy to the right service without this host ?

**Comment by hbagdi** [comment]
> So I apparently can't use the host preperty in my kong-ingress the redicecrt the url to the right backend service. So how can redirect those request to the kong-proxy to the right service without this host ?

You can do path based routing where you can configure traffic for a path starting with, for example, `/foo` can be forwarded to a specific service. It is possible to accomplish this by creating an Ingress resource without a `host` attribute in the rules.

**Comment by samsja** [comment]
Yes, but in this case it is harder to route the traffic to the right service ! I need to manage many APIs that start for exemple by /api/v1/compute, how do i specify that I want to compute on api-A and not api-B that have the exact same route.

Moreover: with this solution 

```
  - http:
      paths:
      - backend:
          serviceName: back-end-service
          servicePort: 80
        path: /api/v1/
```

when I try to reach the back-end API with https://LOAD_BALANCER_IP/api/v1/ it work and I reach the route /api/v1/ of the api.

However when I want try to compute with https://LOAD_BALANCER_IP/api/v1/ compute the ingress ressources redirect to the api but I only reach the /api/v1/ endpoint and note the /api/v1/compute one.

In this case Host based routing seems to be the solution to avoid all this issues but because of the CORS limitation I can't proprely expose my APIs...

**Comment by hbagdi** [comment]
> However when I want try to compute with https://LOAD_BALANCER_IP/api/v1/ compute the ingress ressources redirect to the api but I only reach the /api/v1/ endpoint and note the /api/v1/compute one.

There is probably a configuration error in that case. Please share your Ingress (and the KongIngress) resource in this case.

**Comment by samsja** [comment]
How to reproducde the error:

* install keycloak :
```shell
helm install --name keycloak  --namespace default  stable/keycloak
```
* install kong 
```shell
kubectl apply -f https://bit.ly/kong-ingress
```

* deploy the ingress:

```yaml


apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-app-ingress
  namespace: default
  annotations:
    kubernetes.io/ingress.class: "kong"
    nginx.ingress.kubernetes.io/use-regex: "true"
    certmanager.k8s.io/cluster-issuer: letsencrypt-production-issuer
spec:
  rules:
    - http:
      paths:
      - backend:
          serviceName: keycloak-http
          servicePort: http
        path: /auth/.*
  tls:
  - hosts:
    - keycloak.my-app.mydomaim.com
    secretName: my-app-certificate-tls
```

** issues:

when i want to acces to keycloak main page via : https://keycloak.my-app.mydomaim.com/auth/
( the keycloak.my-app.mydomaim point to kong proxy -ip thanks to my dns)
however when I want to go to the administration console I need to go to the route /auth/realms/master of the keycloak-http service, but when i try to acces it via
https://keycloak.my-app.mydomaim.com/auth/realms/master  
I steel acces the keycloak main page, i reach the /auth/ route of the api and not the /auth/realm/master

The only solution I have found is to change the 
```yaml
        path: /auth/.*
```
to

```yaml
        path:
```

that way I can acces to the right route with the right url, however it is not scalable and I can't add a second backend service with the same ```path: ``` because without the host the ingress ressources won't be able te distinct the two services.

**Comment by bungle** [comment]
Umm.. why not have one route with `paths: /auth` and other with `paths: /auth/realms`? You can then have different service on each?

**Comment by samsja** [comment]
Yes ouf course I can but I am some how limited because I will need to repath EVERY route that my api dispose, and in the case of keycloak I don't even know how many routes they are, but I am sure that they are a lot. I need to manage a lot of api so I am pretty sure that your solution is not scalable even if it will work. That is why I hope to find something close to the "host" option in the ingress, or perhaps a way to cheat and avoid CROS error. Thanks !
